MazerPlatformer/GameWorld.cs

```diff
@@ -145,61 +148,88 @@ public void Initialize()
-                thePlayer.UpdateComponentByType(Component.ComponentType.Health, newHealth);
-
-                if (newHealth <= 0)
-                {
-                    _level.PlayLoseSound();
-                    _playerDied = true;
-                    OnPlayerDied?.Invoke(thePlayer.Components);
-                }
+                return from hitPointsComponent in otherObject1.FindComponentByType(Component.ComponentType.HitPoints)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    from healthComponent in gameObject.FindComponentByType(Component.ComponentType.Health)
+                        .ToEither(NotFound.Create("Could not find health component"))
+                    let myHealth = (int) healthComponent.Value
+                    let hitPoints = (int) hitPointsComponent.Value
+                    select myHealth - hitPoints;
             }
 
-            if (npcType == Npc.NpcTypes.Pickup)
+            Either<IFailure, int> DetermineNewLevelPoints(GameObject thePlayer1, GameObject gameObject1)
             {
-                // pickup points
-                var pickupPoints = (int)otherObject.FindComponentByType(Component.ComponentType.Points).Value;
-                var myPoints = (int)thePlayer.FindComponentByType(Component.ComponentType.Points).Value;
-                var levelPoints = myPoints + pickupPoints;
-                thePlayer.UpdateComponentByType(Component.ComponentType.Points, levelPoints);
+                return from pickupPointsComponent in gameObject1.FindComponentByType(Component.ComponentType.Points)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    from myPointsComponent in thePlayer1.FindComponentByType(Component.ComponentType.Points)
+                        .ToEither(NotFound.Create("Could not find hit-point component"))
+                    let myPoints = (int) myPointsComponent.Value
+                    let pickupPoints = (int) pickupPointsComponent.Value
+                    select myPoints + pickupPoints;
             }
         }
 
+        
+
 
         /// <summary>
         /// We ask each game object within the game world to draw itself
         /// </summary>
         /// <param name="spriteBatch"></param>
-        public void Draw(SpriteBatch spriteBatch)
-        {
-            if (_unloading) return;
-            foreach (var gameObject in _gameObjects.Values.Where(obj => obj.Active))
-            {
-                if (gameObject.Active)
-                    gameObject.Draw(spriteBatch);
-            }
-        }
+        public Either<IFailure, Unit> Draw(SpriteBatch spriteBatch)
+            => _unloading ? Nothing
+                          : _gameObjects.Values
+                              .Where(obj => obj.Active)
+                              .Select(gameObject => gameObject.Draw(spriteBatch))
+                              .AggregateFailures();
```

Comment:
Generally, looking at this code I wonder if I could have written this in the way it looks now...
There is definitely a way of thinking that needs to change in order to first think like this...and not do this post-production.